experiments/kompanion/cmd/migrator/migrator.go

```diff
@@ -0,0 +1,369 @@
+	if err != nil {
+		return fmt.Errorf("error fetching namespaces: %w", err)
+	}
+	shouldExcludeObject := func(id types.NamespacedName) bool {
+		if shouldExclude(id.Namespace, opts.IgnoreNamespaces, opts.TargetNamespaces) {
+			return true
+		}
+		return shouldExclude(id.Name, opts.IgnoreObjects, opts.TargetObjects)
+	}
+
+	// write the namespaces to the migrator file
+	filedir := filepath.Dir(opts.MigrationFile)
+	if err := os.MkdirAll(filedir, 0o700); err != nil {
+		return fmt.Errorf("error creating directory %s: %w", filedir, err)
+	}
+	migfile, err := os.Create(opts.MigrationFile)
+	if err != nil {
+		return fmt.Errorf("error creating migration file %s: %w", opts.MigrationFile, err)
+	}
+	defer migfile.Close()
+	nsGvr := schema.GroupVersionResource{
+		Group:    "",
+		Version:  "v1",
+		Resource: "namespaces",
+	}
+	nss, err := dynamicClient.Resource(nsGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("fetching namespace %s resources: %w", nsGvr, err)
+	}
+	for _, ns := range nss.Items {
+		ns = pruneResource(ns)
+		if len(ns.GetAnnotations()) <= 0 {
+			continue
+		}
+		data, err := yaml.Marshal(ns)
+		if err != nil {
+			return fmt.Errorf("error marshalling namespace: %w", err)
+		}
+		migfile.WriteString(string(data[:]))
+		migfile.WriteString("---\n")
+	}
+
+	// create the work log for go routine workers to use
+	q := &taskQueue{}
+
+	// Parallize across resources, unless we are scoped to a few namespaces
+	// The thought is that if users target a particular namespace (or a few), they may not have cluster-wide permission.
+	perNamespace := len(opts.TargetNamespaces) > 0
+	if perNamespace {
+		for _, ns := range namespaces.Items {
+			if shouldExclude(ns.Name, opts.IgnoreNamespaces, opts.TargetNamespaces) {
+				continue
+			}
+
+			q.AddTask(&dumpResourcesTask{
+				Namespace:     ns.Name,
+				Resources:     resources,
+				DynamicClient: dynamicClient,
+				OutputChannel: outputChannel,
+				// ReportDir:           reportTempDir,
+				ShouldExcludeObject: shouldExcludeObject,
+			})
+		}
+	} else {
+		for _, resource := range resources {
+			q.AddTask(&dumpResourcesTask{
+				Resources:     []schema.GroupVersionResource{resource},
+				DynamicClient: dynamicClient,
+				OutputChannel: outputChannel,
+				// ReportDir:           reportTempDir,
+				ShouldExcludeObject: shouldExcludeObject,
+			})
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	var errs []error
+	var errsMutex sync.Mutex
+
+	for i := 0; i < opts.WorkerRoutines; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				task := q.GetWork()
+				if task == nil {
+					// no work
+					return
+				}
+
+				if err := task.Run(ctx); err != nil {
+					errsMutex.Lock()
+					errs = append(errs, err)
+					errsMutex.Unlock()
+				}
+			}
+		}()
+	}
+
+	go func() {
+		for data := range outputChannel {
+			migfile.WriteString(data)
+			migfile.WriteString("---\n")
+		}
+	}()
```

Comment:
For larger clusters it would be really cool if we had some progress bar or some other kind of progress reporting -- just a thought